pkg/controllers/repositoryserver/repositoryserver_controller.go

```diff
@@ -71,37 +71,66 @@ func (r *RepositoryServerReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		return ctrl.Result{}, errors.Wrap(err, "Failed to get a k8s client")
 	}
 
-	logger.Info("Fetch RepositoryServer CR. If not found end reconcile loop")
 	repositoryServer := &crkanisteriov1alpha1.RepositoryServer{}
 	if err = r.Get(ctx, req.NamespacedName, repositoryServer); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	logger.Info("Setting the CR status as 'Pending' since a create or update event is in progress")
 	repositoryServer.Status.Progress = crkanisteriov1alpha1.Pending
 
-	logger.Info("Found RepositoryServer CR. Create or update owned resources")
 	repoServerHandler := newRepositoryServerHandler(ctx, req, logger, r, kubeCli, repositoryServer)
 	repoServerHandler.RepositoryServer = repositoryServer
 
 	if err = r.Status().Update(ctx, repoServerHandler.RepositoryServer); err != nil {
 		return ctrl.Result{}, err
 	}
+
+	logger.Info("Create or update owned resources")
 	if err := repoServerHandler.CreateOrUpdateOwnedResources(ctx); err != nil {
 		logger.Info("Setting the CR status as 'Failed' since an error occurred in create/update event")
-		repoServerHandler.RepositoryServer.Status.Progress = crkanisteriov1alpha1.Failed
-		if uerr := r.Status().Update(ctx, repoServerHandler.RepositoryServer); uerr != nil {
+		if uerr := repoServerHandler.updateProgressInCRStatus(ctx, crkanisteriov1alpha1.Failed); uerr != nil {
 			return ctrl.Result{}, uerr
 		}
 		r.Recorder.Event(repoServerHandler.RepositoryServer, corev1.EventTypeWarning, "Failed", err.Error())
 		return ctrl.Result{}, err
 	}
-	logger.Info("Setting the CR status as 'Ready' after completing the create/update event\n\n\n\n")
-	repoServerHandler.RepositoryServer.Status.Progress = crkanisteriov1alpha1.Ready
-	if err = r.Status().Update(ctx, repoServerHandler.RepositoryServer); err != nil {
+
+	logger.Info("Connect to Kopia Repository")
+	if err := repoServerHandler.connectToKopiaRepository(); err != nil {
+		if uerr := repoServerHandler.updateProgressInCRStatus(ctx, crkanisteriov1alpha1.Failed); uerr != nil {
+			return ctrl.Result{}, uerr
+		}
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Start Kopia Repository Server")
+	if err := repoServerHandler.startRepoProxyServer(ctx); err != nil {
+		if uerr := repoServerHandler.updateProgressInCRStatus(ctx, crkanisteriov1alpha1.Failed); uerr != nil {
+			return ctrl.Result{}, uerr
+		}
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Add/Update users")
+	if err := repoServerHandler.createOrUpdateClientUsers(ctx); err != nil {
+		if uerr := repoServerHandler.updateProgressInCRStatus(ctx, crkanisteriov1alpha1.Failed); uerr != nil {
+			return ctrl.Result{}, uerr
+		}
 		return ctrl.Result{}, err
 	}
 
+	logger.Info("Refresh Server")
+	if err := repoServerHandler.refreshServer(ctx); err != nil {
+		if uerr := repoServerHandler.updateProgressInCRStatus(ctx, crkanisteriov1alpha1.Failed); uerr != nil {
+			return ctrl.Result{}, uerr
+		}
+		return ctrl.Result{}, err
+	}
+
```

Comment:
Can we move all the kopia setup-related code to a separate method to make the functions smaller?

Reply:
I did not get it clearly, the above separation are based on the logical division we have in the steps of reconciliation. Do you mean combining some of them.

Reply:
The function has become too large. We can move the kopia setup related functions to a separate function. Ideally the functions should be as small as possible

Reply:
ok, I think we can do that.

Reply:
@PrasadG193 I have addressed the comments please check.